### ISSUE TITLE
Windows 10 SDK port to Python

### ISSUE THREAD

**Issue by kdschlosser** [nvaccess/nvda#9278]
I have spent 100's of hours making as close to an exact replica of the Windows 10 SDK written in pure Python  (in it's entirety). I am not yet finished with it but I though it might be of use to this project. At the very least it will be a pretty large time saver if adding any new ctypes or comtypes code you can copy the python code that is already written from the library below. 

https://github.com/kdschlosser/pyWinAPI

let me know if there is something that is not currently added. I have written a code converter that will convert .h and .idl files in the Windows SDK into python. It does a pretty decent job, between 90 and 95% accurate. If i specify a .h file to convert it will follow all includes in the file and convert those as well, after proofing the file and making any needed correction the total time spent ranges from several minutes to 2 days. On occasion it may be longer because I will add to the code converter to handle the programming style of a developer I have not come across that has worked on the Windows SDK (and there are a heap of them)..

The code converter outputs the files exactly as you see in the library. exact same formatting. I never much liked the output of the comtypes code generator. (nor it's speed). I also believe this library has the most complete set of windows data type conversions available.

**Comment by DrSooom** [comment]
@kdschlosser: Please change the issue title to something with more substance. Thanks. 😉

**Comment by Adriani90** [comment]
cc: @michaelDCurran, @jcsteh, @josephsl, @leonardder

**Comment by josephsl** [comment]
Hi there,

Thanks for an interest in making NVDA better through use of Python bindings for Windows 10 SDK.

In addition to Daniel's suggestion, I'd like to request a favor from you. Please do all steps listed below and report back with your findings:

1. Download and compile NVDA from source.
2. Try using certain Windows 10 functions (such as OneCore synthesizer support) with current approach versus the library you're proposing.

Then answer all of the questions below:

1. Can you think of a situation in NVDA that your library can come in and help?
2. Which Windows 10 SDK build does this library cover?
3. Do you observe any issues such as performance differences when you use the original approach as outlined in NVDA source code versus your library?

Thanks very much.

**Comment by josephsl** [comment]
Hi everyone,

To @kdschlosser: this is Joseph, one of the Windows 10 SDK people for NVDA community (the other is @MichaelDCurran). I'm sorry to say that your library might not be helpful for us at this time. After looking at your repo, I'm afraid the header we may wish to look at isn't there at all: UIAutomation*.idl is missing. UIAutomationCore.dll is the library NVDA uses to communicate with UI Automation controls, and without looking at Python version of that, we won't be able to ascertain if the library can even be used. Note that it isn't just NVDA that requires this library - any Python app wishing to use UIA (either as a server and/or a client) may need to include that missing file.

To others: based on git log of Kevin's repo, it looks to me the latest SDK build is 18317, which may or may not be useful for NVDA. Due to change being constant in Windows 10 world, I'd say the cut-off build should be 17134 (or better yet, build 17763 materials such as IUIAutomation6 must be present).

I'm keeping this issue open until Kevin gives answers to questions I posed above, and then we'll think about leaving it open or closing it for now. Thanks.

**Comment by kdschlosser** [comment]
The missing component you speak of I can easily create. I wanted to show you the repo. not necessarily to use as a whole. I browsed through the source code for NVDA and had noticed quite a few ctypes structures for use with Windows. I was simply recommending a reference or perhaps using to "copy and paste" from instead of making the conversions yourself by hand, a time saver.  I am pretty sure the code I wrote is based on the Windows 10 SDK version 10.0.17763.0. When i looked up the SDK on Microsofts website it bring up this version, the VS 2017 installer also does not show a version that is newer then this one. So I am not sure on what version is the latest. UI know that I am using the latest one that is shown in thee VS 2017 installer which is thee version above.

As you know the Windows 10 SDK does not only apply to Windows 10. It applies to all versions of Windows down to XP I believe. The code layout is as close to an exact mirror of the code written in the SDK. Because of this it handles creating the correct objects based on the Windows version and processor architecture. It also makes it really easy to find information about different aspects of it. So if you go onto the Windows Development Knowledgebase and search for IMMNotificationClient the information you are going to see will pertain to that COM object. the knowledge base is also going to tell you that the h file containing the API for the COM object is located in mmdeviceapi.h. Because of the mirrored structure it is in the exact same location in the pyWinApi library. \um\mmdeviceapi_h

Any code written in C/C++ is going to be faster then code written in python. This would be more of a reference for the parts and pieces of Windows that you want to access that may not be worth the time to write and compile DLL's for. I have personally found the performance impact negligible when accessing any of the Windows API using ctypes and Python. The only places where I have seen large performance hits are if any complex math is involved. Would I use this library to custom draw a GUI for a program.....No.. Even tho it could be used to do that. Using the crypto API works fine because all of the complex math is handled in C/C++ code by Windows. 

My post was not intended to have the library included into this project. It is not finished. But if there was perhaps a new feature you wanted to add to NVDA and wanted to save some time and not have to write the ctypes objects it would be worth taking a look see if I have that part converted. Even if I do not have it converted you can ask and in most cases I can get it done in very short order.

**Comment by josephsl** [comment]
Hi, thanks for the detailed answer. @MichaelDCurran, can we perhaps collaborate with this person in some form? Thanks.

**Comment by kdschlosser** [comment]
I did the conversion on the UIAutomation h and idl files. I added them to the pyWinAPI repository

**Comment by josephsl** [comment]
Hi, yep, I see it, and the version I’m seeing is based on 10.0.17134; IUIAutomation6 has been added in 10.0.17763, which is the version NVDA internally uses. Thanks.

**Comment by kdschlosser** [comment]
ok then It looks as tho I will have to update my SDK version.

I will write a small program that will compare the 2 SDK versions for any differences then populate those differences into my library.

**Comment by kdschlosser** [comment]
I know this is kind of sort of another topic... But I had an idea... I think you may like this one a whole lot.
I am the Administrator of a project that I am pretty sure we could work together on this.

I do not know the full scope of NVDA. but from what I have assessed from reading the website is this. Its an accessibility package for the visually impaired that runs on Microsoft Windows. It aids in identification of various components, GUI objects.. text.. things of that nature. This is obviously the short an dirty description. I am sure it is much more involved. I think this is a fantastic thing.. and the fact that it's open source is what is truly awesome. So I tip my hat to the folks that have/do/will work on this project. There is a single limitation with the software I think. It only will help with using a Computer... the world is much larger then just a PC. There is a whole plethora of other electronic devices that could benefit from this project.

The project that I head up is called EventGhost. You may or may not have heard of it. It runs on Microsoft Windows as well. The best explanation I can come up with is this. EventGhost is a centralized API that is used to allow 2 different devices to be able to communicate with one another even if they "speak" 2 different languages. You could call it a translator. but it really isn't that either. EventGhost's whole design is around extensibility and a very simple to use API. It gives the user the ability to accomplish tasks using an event -> action design. easy example is: if "This" happens then do "That" this type of logic can be applied to almost everything we do in our lives. and it also easily applies to electronics. EventGhost currently has over 400 plugins. these plugins range from Sony TV's to Xbox controllers. Z-Wave devices to interfacing with MQTT or IFTTT. we are able to get state changes from electronic devices. Even location awareness with a precision between 6 inches and 2 feet elevation included (location awareness inside of a home without the use of motion sensors). If the 2 pieces of software shared information NVDA just became able to help in ways I can't even imagine. It is no longer bound to only helping to operate a PC.

Here would be a scenario.. With the wonderful world of Smart TV's and applications on these TV's like Netflix How would a visually impaired person be able to browse through the available content on a smart TV?. If NVDA and EventGhost shared information this sort of thing can be done very easily. 

EventGhost is also the best piece of software made to handle IR remote controls. Here is a great example. if you have 3 TV's in your home and each TV is made by a different manufacturer either you would have to use universal remotes (not all functions of a TV work from the universal remotes) or memorize each of the TV's buttons. Now think of this scenario (I have this personally set up in my house) all the remotes are from the same company. they are all the same model. and you can walk from room to room with it and it will automatically change to controlling the devices in the room you are in. You don't have to do a thing.

NVDA could even be set up to perform tasks based on button presses of any IR remote. I did say any there. You do not have to use a Windows MCE remote. Everyone has one of those boxes or drawers that has the remotes from devices they no longer have. It will work with any IR remote.

another use example. a user is sitting on the sofa and decides they want to send an e-mail. press a button on the remote and this will cause EventGhost to generate an event.. and that event we have married to an action that will tell NVDA to perform whatever tasks are needed to send an e-mail.

EventGhost is not just geared for communications with devices. It also will control and get information from other software. start programs, stop them. window control (minimize, maximize, focus, show, hide...) events for when a program starts or stops. when it comes into focus.

I do not think I could even give a complete list of all of the things that EventGhost is able to do.
The single most powerful feature of EventGhost is it's real time scripting support.

I did also want to mention that it has a memory footprint of 45ish megs (not a mistake there) and the disk footprint is roughly 85 meg (no mistake there either). It is 98% written in Python and the other 2% is C extensions. 

This is a quote from one of the users of EventGhost.
> EventGhost is quite possibly the only program that will allow you to open your garage door with an XBox controller. 

EventGhost is completely FREE. the software and support. it is also Open Source

http://www.eventghost.net
https://github.com/EventGhost/EventGhost

I know this is off topic. I figured since I was sharing. I would mention it. I was speaking with my wife about NVDA and she said to me
> If you think about the 2 programs fundamentally do the same thing but with different input and output. 
I started thinking.... hmmm... how cool would that be if the 2 programs worked together. With the combined goal of making almost ANY device accessible by anyone that uses the software...

Me nor the people that work on the EventGhost project are interested in any kind of personal gain other then the happiness they get when the users express there excitement/gratitude. I don't know about you but that is one feeling I enjoy having.

If you want me to open a separate issue for the above I can do that.

**Comment by DataTriny** [comment]
Hello @kdschlosser ,
Regarding EventGhost, NVDA has an add-on system that let's us extends its features simply by installing packaged Python code. Since what you are talking about is, I think, out of scope for NVDA core, the solution would be to create an add-on for your software. Don't forget however that NVDA is tightly coupled to Windows, and running it on embedded systems will not be possible. Would like to see if NVDA can communicate with your product as IoT is growing more and more, and it's not the most friendly place for visually impaired people.

**Comment by kdschlosser** [comment]
if NVDA has some kind of an API then EventGhost will be able to communicate with it.

EventGhost is written in python but it does have an executable. So adding it as a library would not be possible. it would have to be able to communicate with NVDA over some kind of connection.

EventGhost is just like NVDA it to is tightly coupled with Windows. And it will not run on any kind of an embedded system.

So what I am thinking is this. if the API is there and there is some kind of ability built into NVDA that exposes this API over some form of a connection. whether it be a named pipe. websockets, REST, SOAP, direct socket to name a few. then it is possible to have to 2 integrate with one another.

so the question would be. is there an exposed API?

**Comment by Andre9642** [comment]
First, thanks for your initial work concerning Windows 10 SDK!
I just tested EventGhost quickly. It seems accessible. You use WXWidget and it is a good thing regarding accessibility with screen readers.
NVDA has an API, [see here](https://github.com/nvaccess/nvda/tree/master/extras/controllerClient). However, note that it's not necessary (and preferable) if your app is natively accessible.

**Comment by kdschlosser** [comment]
define 
> natively accessible.

**Comment by josephsl** [comment]
Hi,

Natively accessible: although it depends on who you ask, I usually define it as an app that is usable by people with disabilities via an assistive technology (such as screen readers) out of the box.

I think, for sake of space, it would be great for you to subscribe to NVDA development mailing list (hosted on SourceForge) so we can have a deeper discussion about accessibility in general on that email list. I can tell you wish to help many people, including possibly people with disabilities; it is just that GitHub might not be a good venue for a really deep discussion on accessibility under the surface, which will open up all sorts of topics for discussions.

Thanks.

**Comment by leonardder** [comment]
@kdschlosser: Are you still working on this project? Must be pretty hard to maintain with all the updates created by Microsoft.

**Comment by kdschlosser** [comment]
Microsoft is making me coo coo.. The issue is not the changes. it's the changes they are making and mentioning or documenting. or they are stating they are not making a change and in fact they do.. It makes it very hard.

I am still very interested in making this happen. There is an enormity of benefits. I have been super busy with working on my house and also further extending EventGhosts capabilities. Adding ZWave, and Security Camera control with facial recognition. These are devices that are natively not accessible and I do not believe there is anyone offering accessibility for devices like this. with NVDA and EventGhost it becomes possible. The security camera thing would be great so it can announce who is at the door.

I did sign up for the mailing list I think.. https://groups.io/g/nvda-devel. if that is the correct place should I open a new topic??

Oh I also added support for Windows 10 Virtual Desktops as well. I personally do not run Windows 10. But from what I have been told is that the desktops are not persistent between restarts... With EventGhost able to control them they are now persistent
